util/text/text.go

```diff
@@ -16,98 +22,241 @@ func Wrap(text string, lineWidth int) (string, int) {
+// this since this function is not able to check it.
+func applyTermEscapes(line string, escapes []EscapeItem) string {
+	if len(escapes) == 0 {
+		return line
+	}
+
+	var out string = ""
+
+	currPos := 0
+	currItem := 0
+	for _, r := range line {
+		if currItem < len(escapes) && currPos == escapes[currItem].pos {
+			// NOTE: We avoid terminal escapes at the end of a line by move them one
+			// pass the end of line, so that algorithms who trim right spaces are
+			// happy. But algorithms who trim left spaces are still unhappy.
+			if r == '\n' {
+				out += "\n" + escapes[currItem].item
 			} else {
-				// Break a word longer than a line
-				if wordLength > lineWidth {
-					for wordLength > 0 && wordLen(word) > 0 {
-						l := minInt(spaceLeft, wordLength)
-						part, leftover := splitWord(word, l)
-						word = leftover
-						wordLength = wordLen(word)
-
-						lineBuffer.WriteString(part)
-						textBuffer.WriteString(pad)
-						textBuffer.Write(lineBuffer.Bytes())
-						lineBuffer.Reset()
-
-						spaceLeft -= l
-
-						if spaceLeft <= 0 {
-							textBuffer.WriteString("\n")
-							nbLine++
-							spaceLeft = lineWidth - leftPad
-						}
-
-						if wordLength <= 0 {
-							break
-						}
-					}
-				} else {
-					// Normal break
-					textBuffer.WriteString(pad + strings.TrimRight(lineBuffer.String(), " "))
-					textBuffer.WriteString("\n")
-					lineBuffer.Reset()
-					lineBuffer.WriteString(word)
-					firstWord = false
-					spaceLeft = lineWidth - leftPad - wordLength
-					nbLine++
-				}
+				out += escapes[currItem].item + string(r)
+				currPos++
 			}
+			currItem++
+		} else {
+			if r != '\n' {
+				currPos++
+			}
+			out += string(r)
 		}
+	}
 
-		if lineBuffer.Len() > 0 {
-			textBuffer.WriteString(pad + strings.TrimRight(lineBuffer.String(), " "))
-			lineBuffer.Reset()
-		}
+	return out
+}
+
+// Segment a line into chunks, where each chunk consists of chars with the same
+// type and is not breakable.
+func segmentLine(s string) []string {
+	var chunks []string
+
+	var word string
+	wordType := NONE
+	flushWord := func() {
+		chunks = append(chunks, word)
+		word = ""
+		wordType = NONE
+	}
 
-		firstLine = false
+	for _, r := range s {
+		// A WIDE_CHAR itself constitutes a chunk.
+		thisType := runeType(r)
+		if thisType == WIDE_CHAR {
+			if wordType != NONE {
+				flushWord()
+			}
+			chunks = append(chunks, string(r))
+			continue
+		}
+		// Other type of chunks starts with a char of that type, and ends with a
+		// char with different type or end of string.
+		if thisType != wordType {
+			if wordType != NONE {
+				flushWord()
+			}
+			word = string(r)
+			wordType = thisType
+		} else {
+			word += string(r)
+		}
+	}
+	if word != "" {
+		flushWord()
 	}
 
-	return textBuffer.String(), nbLine
+	return chunks
+}
+
+// Rune categories
+//
+// These categories are so defined that each category forms a non-breakable
+// chunk. It IS NOT the same as unicode code point categories.
+//
+const (
```

Comment:
There is a few things that could be improved here:
- These are useful only inside the package, so they should not be exported
- CamelCase is generally used in go
- iota could be better used
- an explicit type could be used

It could be something like this:

```go
const (
	none int = -1 + iota
	wideChar
	invisible
	shortUnicode
	space
	visibleAscii
)
```

You probably don't need the `-1` but I'll let you decide.

Reply:
Thanks for the suggestions. I will improve this.